src/data_source/fetching.rs

```diff
@@ -0,0 +1,885 @@
+
+        // We `join_all` here because we want this iterator to be evaluated eagerly for two reasons:
+        // 1. It borrows from `self`, which is local to this future. This avoids having to clone
+        //    `self` for every entry, instead we clone it for every chunk.
+        // 2. We evaluate all the `some_or_fetch` calls eagerly, so the fetches are triggered as
+        //    soon as we evaluate the chunk. This ensures we don't miss any notifications, since we
+        //    load from storage and subscribe to notifications for missing objects all while we have
+        //    a read lock on `self.storage`. No notifications can be sent during this time since
+        //    sending a notification requires a write lock.
+        stream::iter(join_all(ts).await)
+    }
+
+    async fn ok_or_fetch<R, T>(
+        self: &Arc<Self>,
+        storage: &RwLockReadGuard<'_, NotifyStorage<Types, S>>,
+        req: R,
+        res: QueryResult<T>,
+    ) -> Fetch<T>
+    where
+        R: Into<T::Request> + Send,
+        T: Fetchable<Types> + Send + Sync + 'static,
+    {
+        self.some_or_fetch(storage, req, res.ok_or_trace()).await
+    }
+
+    async fn some_or_fetch<R, T>(
+        self: &Arc<Self>,
+        storage: &RwLockReadGuard<'_, NotifyStorage<Types, S>>,
+        req: R,
+        res: Option<T>,
+    ) -> Fetch<T>
+    where
+        R: Into<T::Request> + Send,
+        T: Fetchable<Types> + Send + Sync + 'static,
+    {
+        match res {
+            Some(t) => Fetch::Ready(t),
+            None => self.fetch(storage, req).await,
+        }
+    }
+
+    async fn fetch<R, T>(
+        self: &Arc<Self>,
+        storage: &RwLockReadGuard<'_, NotifyStorage<Types, S>>,
+        req: R,
+    ) -> Fetch<T>
+    where
+        R: Into<T::Request>,
+        T: Fetchable<Types>,
+    {
+        let req = req.into();
+        tracing::debug!("fetching resource {req:?}");
+
+        // Subscribe to notifications so we are alerted when we get the resource.
+        let fut = T::passive_fetch(&**storage, req)
+            .await
+            .then(move |opt| async move {
+                match opt {
+                    Some(t) => t,
+                    None => {
+                        // If `passive_fetch` returns `None`, it means the notifier was dropped
+                        // without ever sending a notification. In this case, the correct behavior
+                        // is actually to block forever (unless the `Fetch` itself is dropped),
+                        // since the semantics of `Fetch` are to never fail. This is analogous to
+                        // fetching an object which doesn't actually exist: the `Fetch` will never
+                        // return.
+                        tracing::warn!(
+                            "notifier dropped without satisfying request {req:?}, fetch will block"
+                        );
+                        future::pending().await
```

Comment:
Hmm this feels like strange behavior - I understand the notion of blocking forever to be consistent with the semantics of fetch, but is there ever a practical reason to block forever in this case?

Reply:
Yeah you're right, panic is probably better here